### DISCUSSION TITLE
Transaction Manager in Sequelize

### DISCUSSION THREAD

**Discussion by Mamun724** [sequelize/sequelize#15915] [Help & Questions] [answered]
We can use the transactions on a single database as documented in the Sequelize documentation. **Is there a way to use a single transaction in more than one PostgreSQL database?** Let me explain the scenario:
Let’s say we have two PostgreSQL databases, dbA, and dbB and 2 tables dbA.table and dbB.table. We want to commit dbA.table table only if dbB.table table is updated properly else roll back both. **Is it possible to do all the manipulations under a single transaction?**
We have found that that is possible using a transaction manager as described [here](https://stackoverflow.com/a/65832350) and [here](https://stackoverflow.com/a/59058207).
Now for Sequelize:
- How can we achieve that, in general?
  - Is there any recommended transaction manager for Sequelize?
  - Or can we configure Sequelize in such a way that we can execute the operations of the above scenario under a single transaction?

If the concept of transaction manager is not limited to Sequelize, **is there any solution for NodeJS?**

**Comment by ephys** [comment, marked as answer]
Sequelize does not provide a built-in way to handle transactions on multiple databases, nor one for prepared transactions. This is something you'll need to manage manually using unmanaged transactions.

This is roughly what this could look like

```ts
// sequelize1 & sequelize2 are two different sequelize instances, connected to different databases.
const id = crypto.randomUUID();
let firstTransactionPrepared = false;

const transaction1 = await sequelize1.transaction(); // In Sequelize 7, this method is called unmanagedTransaction()
try {
  const transaction2 = await sequelize2.transaction();
  try {
    // do work on both databases here
    // ...

    // Prepare & commit
    await sequelize1.query('PREPARE TRANSACTION :id', {
      replacements: { id },
      transaction: transaction1,
    });
    firstTransactionPrepared = true;

    await transaction2.commit();

    await sequelize1.query('COMMIT PREPARED :id', {
      replacements: { id },
      transaction: transaction1,
    });
  } catch (error) {
    await transaction2.rollback();
    throw error;
  }
} catch (error) {
  if (firstTransactionPrepared) {
    await sequelize1.query('ROLLBACK PREPARED :id', {
      replacements: { id },
      transaction: transaction1,
    });
  } else {
    await transaction1.rollback();
  }

  throw error;
}
```

**Reply by Mamun724** [reply]
@ephys, Thank you for your reply. We will get back after some R&D with this approach.

**Reply by Mamun724** [reply]
Hello @ephys,
We have implemented this approach with Sequelize as you suggested.
Today we have noticed that without restarting the node server after performing 5 operations with this code snippet, we are getting a 'connection acquire timeout' error in Sequelize. After some R&D we have found that the process created to execute the query by the sequelize1 is not terminated automatically and thus the connection to PostgreSQL server is not closed automatically. Once we stop the Nodejs server the processes close and the connections also get terminated(We have used PG Admin 4's database dashboard to check if the connections get disconnected). As we were using the default connection pool size(that is 5) for Sequelize, that's why the exception was coming after 5 operations. We can increase the connection pool size but that is also limited and after reaching the limit the server should be restarted. **Now my question is how to automatically close the connection used for the prepared transaction. Is there any other way to close the connection manually after the operation has completed/failed?** Thanks in advance.
